gpflow/posteriors.py

```diff
@@ -292,6 +293,118 @@ def _conditional_fused(
+        self,
+        kernel,
+        X_data: tf.Tensor,
+        Y_data: tf.Tensor,
+        inducing_variable: InducingPoints,
+        likelihood_variance: Parameter,
+        num_latent_gps:int,
+        mean_function: Optional[mean_functions.MeanFunction] = None,
+        *,
+        precompute_cache: Optional[PrecomputeCacheType],
+    ):
+        super().__init__(kernel, X_data, mean_function=mean_function)
+        self.mean_function = mean_function
+        self.X_data = X_data
+        self.Y_data = Y_data
+        self.likelihood_variance = likelihood_variance
+        self.inducing_variable = inducing_variable
+        self.num_latent_gps = num_latent_gps
+
+        if precompute_cache is not None:
+            self.update_cache(precompute_cache)
+
+
+    def _precompute(self) -> Tuple[tf.Tensor, tf.Tensor]:
+
+        """
+        Precomputes the cholesky decomposition of Kmm_plus_s for later reuse we will call
+        base_conditional_with_lm implementation ('Qinv' in the Abstract Posterior class). We also
+        precompute the less compute intensive error term ('alpha' in the Abstract Posterior class)
+        """
+
+        # taken directly from the deprecated SGPR implementation
+        num_inducing = self.inducing_variable.num_inducing
+        err = self.Y_data - self.mean_function(self.X_data)
+        kuf = Kuf(self.inducing_variable, self.kernel, self.X_data)
+        kuu = Kuu(self.inducing_variable, self.kernel, jitter=default_jitter())
+        sigma = tf.sqrt(self.likelihood_variance)
+        L = tf.linalg.cholesky(kuu) # cache alpha, qinv
+        A = tf.linalg.triangular_solve(L, kuf, lower=True) / sigma
+        B = tf.linalg.matmul(A, A, transpose_b=True) + tf.eye(num_inducing, dtype=default_float()) # cache qinv
+        LB = tf.linalg.cholesky(B) # cache alpha
+        Aerr = tf.linalg.matmul(A, err)
+        c = tf.linalg.triangular_solve(LB, Aerr, lower=True) / sigma # cache alpha
+
+        # get intermediate variables
+        Linv = tf.linalg.triangular_solve(L,tf.eye(num_inducing, dtype=default_float()))
+        LBinv = tf.linalg.triangular_solve(LB,tf.eye(num_inducing, dtype=default_float()))
+        Binv = tf.linalg.inv(B) # naive...can do better?
+        tmp = tf.eye(num_inducing, dtype=default_float())-Binv
+
+        # calculate cached values
+        alpha = tf.transpose(Linv) @ tf.transpose(LBinv) @ c
+        Qinv = Linv @ tmp @ Linv
+
+        return alpha, Qinv
+
+    def _conditional_with_precompute(
+        self, Xnew, full_cov: bool = False, full_output_cov: bool = False
+    ) -> MeanAndVariance:
+        """
+        Computes predictive mean and (co)variance at Xnew, *excluding* mean_function.
+        Relies on cached alpha and Qinv.
+        """
+        Kus = Kuf(self.inducing_variable, self.kernel, Xnew)
+        Knn = self.kernel(Xnew, full_cov=full_cov)
+
+        mean = tf.transpose(Kus) @ self.alpha
+        var = Knn - tf.transpose(Kus) @ self.Qinv @ Kus
+
```

Comment:
This is not working. It is also needs to be implemented for ` full_cov: bool = False, full_output_cov: bool = False`

Reply:
@Andrew878, @johnamcleod  I have made the necessary changes to make this work:
- there was an error in that the mean was returned with the mean function values added
- the provided formula for Qinv is incorrect, I have changed the fast predictions notebook accordingly
- I have taken into account the two `full_cov` cases

The tests pass for all `full_cov` cases, but the notebook is not happy in the `full_cov=False` case.

Reply:
@Andrew878 @johnamcleod This is not working in both cases. Note how we don't need to factorise `Qinv` in the `full_cov=False` case any more.